File: bindings/java/java/src/test/java/com/stateset/embedded/CryptoVectorTests.java
package com.stateset.embedded;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Cross-binding compatibility test for the Java JNI binding.
 *
 * <p>Reads the language-neutral corpus at
 * {@code bindings/test-vectors/v1.json} and asserts the Java binding
 * produces byte-equal hex digests to Rust ground truth for every entry.
 * Counterparts: Rust ({@code crates/stateset-crypto/tests/cross_binding_vectors.rs}),
 * Node, Python, Go, and WASM.
 */
class CryptoVectorTests {

    /**
     * The corpus is at workspace-root {@code bindings/test-vectors/v1.json};
     * tests run from {@code bindings/java/java/}, so corpus is at
     * {@code ../../test-vectors/v1.json}.
     */
    private static final Path CORPUS_PATH =
        Paths.get("..", "..", "test-vectors", "v1.json");

    private static final Gson GSON = new Gson();

    private static JsonObject loadCorpus() throws IOException {
        String raw = new String(Files.readAllBytes(CORPUS_PATH), StandardCharsets.UTF_8);
        JsonObject parsed = JsonParser.parseString(raw).getAsJsonObject();
        assertEquals(1, parsed.get("version").getAsInt(), "corpus version must be 1");
        return parsed;
    }

    private static String hex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            sb.append(String.format("%02x", b & 0xff));
        }
        return sb.toString();
    }

    private static byte[] fromHex(String s) {
        int len = s.length();
        byte[] out = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            out[i / 2] = (byte) ((Character.digit(s.charAt(i), 16) << 4)
                + Character.digit(s.charAt(i + 1), 16));
        }
        return out;
    }

    @Test
    void corpusIsPresentAndVersionOne() throws IOException {
        JsonObject f = loadCorpus();
        JsonObject categories = f.getAsJsonObject("categories");
        assertTrue(categories.has("canonical_json"));
        assertTrue(categories.has("payload_plain_hash"));
        assertTrue(categories.has("merkle_root"));
    }

    @Test
    void canonicalJsonVectorsMatchGroundTruth() throws IOException, NoSuchAlgorithmException {
        JsonObject f = loadCorpus();
        JsonArray vectors = f.getAsJsonObject("categories").getAsJsonArray("canonical_json");
        for (JsonElement el : vectors) {
            JsonObject v = el.getAsJsonObject();
            String id = v.get("id").getAsString();
            String input = v.get("input").toString();
            String expected = v.get("expected_hex").getAsString();

            byte[] canonical = Crypto.jcsCanonicalize(input);
            byte[] digest = MessageDigest.getInstance("SHA-256").digest(canonical);
            assertEquals(expected, hex(digest),
                "canonical_json/" + id + ": SHA-256(jcs(input)) mismatch");
        }
    }

    @Test
    void payloadPlainHashVectorsMatchGroundTruth() throws IOException {
        JsonObject f = loadCorpus();
        JsonArray vectors = f.getAsJsonObject("categories").getAsJsonArray("payload_plain_hash");
        for (JsonElement el : vectors) {
            JsonObject v = el.getAsJsonObject();
            String id = v.get("id").getAsString();
            String input = v.get("input").toString();
            String expected = v.get("expected_hex").getAsString();
            byte[] salt = (v.has("salt_hex") && !v.get("salt_hex").isJsonNull())
                ? fromHex(v.get("salt_hex").getAsString())
                : null;

            byte[] digest = Crypto.payloadPlainHash(input, salt);
            assertEquals(expected, hex(digest),
                "payload_plain_hash/" + id + ": digest mismatch");
        }
    }

    @Test
    void merkleRootVectorsMatchGroundTruth() throws IOException {
        JsonObject f = loadCorpus();
        JsonArray vectors = f.getAsJsonObject("categories").getAsJsonArray("merkle_root");
        for (JsonElement el : vectors) {
            JsonObject v = el.getAsJsonObject();
            String id = v.get("id").getAsString();
            JsonArray leavesHex = v.getAsJsonArray("leaves_hex");
            String expected = v.get("expected_hex").getAsString();

            byte[][] leaves = new byte[leavesHex.size()][];
            for (int i = 0; i < leavesHex.size(); i++) {
                leaves[i] = fromHex(leavesHex.get(i).getAsString());
            }
            byte[] root = Crypto.merkleRoot(leaves);
            assertEquals(expected, hex(root),
                "merkle_root/" + id + ": root mismatch");
        }
    }
}

File: bindings/dotnet/tests/CryptoVectorTests.cs
using System.IO;
using System.Security.Cryptography;
using System.Text.Json;
using StateSet.Embedded;
using Xunit;

namespace StateSet.Tests;

/// <summary>
/// Cross-binding compatibility test for the .NET P/Invoke binding.
/// </summary>
/// <remarks>
/// Reads the language-neutral corpus at
/// <c>bindings/test-vectors/v1.json</c> and asserts the .NET binding
/// produces byte-equal hex digests to Rust ground truth for every entry.
/// Counterparts: Rust (<c>crates/stateset-crypto/tests/cross_binding_vectors.rs</c>),
/// Node, Python, Go, WASM, Java, and Kotlin.
/// </remarks>
public class CryptoVectorTests
{
    /// <summary>
    /// The corpus is at workspace-root <c>bindings/test-vectors/v1.json</c>;
    /// xUnit runs from <c>bindings/dotnet/tests/bin/.../</c>, so we walk up
    /// from the current directory until we find <c>bindings/test-vectors/v1.json</c>.
    /// </summary>
    private static string FindCorpus()
    {
        var dir = new DirectoryInfo(Directory.GetCurrentDirectory());
        while (dir != null)
        {
            var candidate = Path.Combine(
                dir.FullName, "bindings", "test-vectors", "v1.json");
            if (File.Exists(candidate)) return candidate;
            dir = dir.Parent;
        }
        throw new FileNotFoundException(
            "could not locate bindings/test-vectors/v1.json from " +
            Directory.GetCurrentDirectory());
    }

    private static JsonDocument LoadCorpus()
    {
        var raw = File.ReadAllText(FindCorpus());
        var doc = JsonDocument.Parse(raw);
        Assert.Equal(1, doc.RootElement.GetProperty("version").GetInt32());
        return doc;
    }

    private static string Hex(byte[] bytes)
    {
        var sb = new System.Text.StringBuilder(bytes.Length * 2);
        foreach (var b in bytes)
        {
            sb.AppendFormat("{0:x2}", b);
        }
        return sb.ToString();
    }

    private static byte[] FromHex(string s)
    {
        var len = s.Length;
        var bytes = new byte[len / 2];
        for (var i = 0; i < len; i += 2)
        {
            bytes[i / 2] = System.Convert.ToByte(s.Substring(i, 2), 16);
        }
        return bytes;
    }

    [Fact]
    public void CorpusIsPresentAndVersionOne()
    {
        using var doc = LoadCorpus();
        var cats = doc.RootElement.GetProperty("categories");
        Assert.Equal(JsonValueKind.Array, cats.GetProperty("canonical_json").ValueKind);
        Assert.Equal(JsonValueKind.Array, cats.GetProperty("payload_plain_hash").ValueKind);
        Assert.Equal(JsonValueKind.Array, cats.GetProperty("merkle_root").ValueKind);
    }

    [Fact]
    public void CanonicalJsonVectorsMatchGroundTruth()
    {
        using var doc = LoadCorpus();
        using var sha = SHA256.Create();
        var vectors = doc.RootElement
            .GetProperty("categories")
            .GetProperty("canonical_json");
        foreach (var v in vectors.EnumerateArray())
        {
            var id = v.GetProperty("id").GetString()!;
            var input = v.GetProperty("input").GetRawText();
            var expected = v.GetProperty("expected_hex").GetString()!;

            var canonical = Crypto.JcsCanonicalize(input);
            var digest = sha.ComputeHash(canonical);
            Assert.Equal(expected, Hex(digest));
            _ = id; // identifies the vector in failure context
        }
    }

    [Fact]
    public void PayloadPlainHashVectorsMatchGroundTruth()
    {
        using var doc = LoadCorpus();
        var vectors = doc.RootElement
            .GetProperty("categories")
            .GetProperty("payload_plain_hash");
        foreach (var v in vectors.EnumerateArray())
        {
            var id = v.GetProperty("id").GetString()!;
            var input = v.GetProperty("input").GetRawText();
            var expected = v.GetProperty("expected_hex").GetString()!;
            byte[]? salt = null;
            if (v.TryGetProperty("salt_hex", out var saltEl) &&
                saltEl.ValueKind == JsonValueKind.String)
            {
                salt = FromHex(saltEl.GetString()!);
            }

            var digest = Crypto.PayloadPlainHash(input, salt);
            Assert.Equal(expected, Hex(digest));
            _ = id;
        }
    }

    [Fact]
    public void MerkleRootVectorsMatchGroundTruth()
    {
        using var doc = LoadCorpus();
        var vectors = doc.RootElement
            .GetProperty("categories")
            .GetProperty("merkle_root");
        foreach (var v in vectors.EnumerateArray())
        {
            var id = v.GetProperty("id").GetString()!;
            var leavesHex = v.GetProperty("leaves_hex");
            var expected = v.GetProperty("expected_hex").GetString()!;

            var leaves = new byte[leavesHex.GetArrayLength()][];
            int i = 0;
            foreach (var h in leavesHex.EnumerateArray())
            {
                leaves[i++] = FromHex(h.GetString()!);
            }
            var root = Crypto.MerkleRoot(leaves);
            Assert.Equal(expected, Hex(root));
            _ = id;
        }
    }
}
